master/internal/command/command_manager.go

```diff
@@ -1,81 +1,112 @@
-		for _, user := range msg.UserIds {
-			userIds[user] = true
-		}
-		for _, command := range ctx.AskAll(&commandv1.Command{}, ctx.Children()...).GetAll() {
-			typed := command.(*commandv1.Command)
-			if (len(users) == 0 && len(userIds) == 0) || users[typed.Username] || userIds[typed.UserId] {
-				resp.Commands = append(resp.Commands, typed)
-			}
+	resp := &apiv1.GetCommandsResponse{}
+	cmds, users, userIDs := cs.listNTSCType(req.Users, req.UserIds, model.TaskTypeCommand)
+	for _, c := range cmds {
+		cmd := c.toCommand()
+		// skip if it doesn't match the requested workspaceID if any.
+		if req.WorkspaceId != 0 && req.WorkspaceId != cmd.WorkspaceId {
+			continue
 		}
-		ctx.Respond(resp)
-
-	case *apiv1.DeleteWorkspaceRequest:
-		ctx.TellAll(msg, ctx.Children()...)
-
-	case *CreateGeneric:
-		taskID := model.NewTaskID()
-		jobID := model.NewJobID()
-		msg.Spec.CommandID = string(taskID)
-		if err := createGenericCommandActor(
-			ctx, c.db, c.rm, taskID, model.TaskTypeCommand, jobID,
-			model.JobTypeCommand, msg.Spec, msg.ContextDirectory,
-		); err != nil {
-			ctx.Log().WithError(err).Error("failed to launch command")
-			ctx.Respond(err)
-		} else {
-			ctx.Respond(taskID)
+		if (len(users) == 0 && len(userIDs) == 0) || users[cmd.Username] || userIDs[cmd.UserId] {
+			resp.Commands = append(resp.Commands, cmd)
 		}
+	}
+	return resp, nil
+}
+
+func (cs *commandService) GetCommand(req *apiv1.GetCommandRequest) (*apiv1.GetCommandResponse, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	c, err := cs.getNTSC(model.TaskID(req.CommandId), model.TaskTypeCommand)
+	if err != nil {
+		return nil, api.NotFoundErrs("command ID", req.CommandId, true)
+	}
+
+	return &apiv1.GetCommandResponse{
+		Command: c.toCommand(),
+		Config:  protoutils.ToStruct(c.Config),
+	}, nil
+}
+
+func (cs *commandService) KillCommand(req *apiv1.KillCommandRequest) (*apiv1.KillCommandResponse, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	c, err := cs.getNTSC(model.TaskID(req.CommandId), model.TaskTypeCommand)
+	if err != nil {
+		return nil, err
+	}
+
+	err = task.DefaultService.Signal(c.allocationID, task.KillAllocation, "user requested kill")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to kill allocation")
+	}
+	return &apiv1.KillCommandResponse{Command: c.toCommand()}, nil
+}
+
+func (cs *commandService) SetCommandPriority(
+	req *apiv1.SetCommandPriorityRequest,
+) (*apiv1.SetCommandPriorityResponse, error) {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+
+	c, err := cs.getNTSC(model.TaskID(req.CommandId), model.TaskTypeCommand)
+	if err != nil {
+		return nil, err
+	}
 
-	case echo.Context:
-		ctx.Respond(echo.NewHTTPError(http.StatusNotFound, api.ErrAPIRemoved))
+	err = c.setNTSCPriority(int(req.Priority), true)
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.SetCommandPriorityResponse{Command: c.toCommand()}, nil
+}
 
-	default:
-		return actor.ErrUnexpectedMessage(ctx)
+func (c *command) toCommand() *commandv1.Command {
```

Comment:
we need at least a read lock over command state here right?

Reply:
I don't remember why I took the lock off -- I think it was also the whole 'private vs public' function thing. maybe I'm following that rule too strictly.